### PR TITLE
Adding Context and RequestOptions support

### PR DESCRIPTION
**Pull request by douglasmiller**
# Using a context with operations

Each non-list operation has had a *WithContext version of the same operation added to the client.

**Note**: Some operations allow the `Context` to be passed in the `Params` of the request. This method is still supported, and the context passed in via `Params` will continue to be used for operations that do not accept the context as the first parameter (operations that end in `WithContext`)

```go
ctx, _ := context.WithTimeout(context.Background(), 20*time.Second)

accountReq := &recurly.AccountCreate{
	...
}

account, err := client.CreateAccountWithContext(ctx, accountReq)
```

Each list operation has had a `FetchWithContext()` and `CountWithContext()` function added in addition to their original non-context based versions.

```go
listParams := &recurly.ListAccountsParams{
	...
}
accounts, err := client.ListAccounts(listParams)

// A single context that applies to all Fetch() calls
ctx, _ := context.WithTimeout(context.Background(), 20*time.Second)
for accounts.HasMore {
	err := accounts.FetchWithContext(ctx)
	if e, ok := err.(*recurly.Error); ok {
		fmt.Printf("Failed to retrieve next page: %v", e)
		break
	}
	for i, account := range accounts.Data {
		fmt.Printf("Account %s fetched\n", account.Code)
	}
}
```

# RequestOptions

Updates client operations to be variadic functions that accept additional request configurations.

Currently there are 2 request options that are supported:
- `WithIdempotencyKey(string)`: Sets the `Idempotency-Key` header to the http request's headers.
- `WithHeader(http.Header)`: Sets additional custom headers to the http request's headers.

**Note**: Some operations allow the `IdempotencyKey` and `Header` values to be passed in the `Params` of the request. This method is still supported, and any values passed in via `Params` will continue to be used. The `With*` values will overwrite existing values provided in `Params`.

**Using `Params`:**
```go
headers := http.Header{"Accept-Language": []string{"fr"}}

accountReq := &recurly.AccountCreate{
	...
	Params: recurly.Params{
		Header: headers,
	},
}

account, err := client.CreateAccount(accountReq)
```

**Using `WithHeader` request option:**
```go
headers := http.Header{"Accept-Language": []string{"fr"}}

accountReq := &recurly.AccountCreate{
	...
}

account, err := client.CreateAccount(accountReq, recurly.WithHeader(headers))
```

### CODE REVIEW

File: params_test.go [hunk 1]
```diff
@@ -20,3 +22,66 @@ func TestFormatTime(test *testing.T) {
 	formattedTime := formatTime(time)
 	t.Assert(formattedTime, formatTime(time), "formatTime()")
 }
+
+func TestParamsWithOptions(test *testing.T) {
+	t := &T{test}
+
+	ctx := context.Background()
+	iKey := "asdf"
+	p := &Params{}
```
Comment (bmacnee) [inline params_test.go:31]: I would say what we want a `New` function for creating params, e.g.
```go
func NewParams(opts ...Option)
```
that way you don't have to call `ApplyOptions` thereafter, the new function can just iterate through them all in one step

Leave the `ApplyOptions` method too, so there's flexibility in when a user decides to add more options

Comment (bmacnee) [reply params_test.go:31]: Also sets us up nicely for creating any defaults the params should have in the future

File: params.go [hunk 2]
```diff
@@ -30,6 +30,58 @@ type Params struct {
 	Data interface{} `json:"-"`
 }
 
+// CommonParams should be used where there are no operation specific Params defined
+type CommonParams struct {
```
Comment (bmacnee) [inline params.go:34]: I'm fuzzy on what this CommonParams vs Params means. CommonParams is just a Params, but then we make a Params pointer when we call toParams()... but CommonParams is already type Params?

File: params.go [hunk 3]
```diff
@@ -30,6 +30,58 @@ type Params struct {
...
+		Context:        list.Context,
+		RequestParams:  list,
+	}
+}
+
+// Option allows for customizations to a request.
+type Option func(p *Params)
+
+// CommonParamsWithOptions will create a CommonParams and apply the passed Options.
+func CommonParamsWithOptions(opts []Option) *CommonParams {
```
Comment (bmacnee) [inline params.go:51]: I like this pattern of just creating the params with the options all at once. let's provide a similar function for just Params, i noted this on the test file for more details

File: params.go [hunk 4]
```diff
@@ -30,6 +30,58 @@ type Params struct {
...
+	c := &CommonParams{}
+	c.Params.ApplyOptions(opts)
+	return c
+}
+
+// ApplyOptions will apply the passed Options to the Params.
+func (p *Params) ApplyOptions(opts []Option) {
```
Comment (bmacnee) [inline params.go:58]: can we keep the signature the same here as in other calls where we take a variadic argument for opts instead of a slice? just for consistency

File: params_test.go [hunk 5]
```diff
@@ -20,3 +22,66 @@ func TestFormatTime(test *testing.T) {
...
+	p.ApplyOptions([]Option{
+		WithIdempotencyKey(iKey),
+		WithContext(ctx),
+	})
+
+	t.Assert(iKey, p.IdempotencyKey, "ApplyOptions()")
+	t.Assert(ctx, p.Context, "ApplyOptions()")
+}
+
+func TestCommonParamsWithOptions(test *testing.T) {
```
Comment (bmacnee) [inline params_test.go:41]: let's add a test case where no options are passed in, to make sure we dont hit any nil pointer issues. here's how i like to set up multiple test cases within one test function. by calling t.Run() it will treat each iteration as a unique test, and it saves you on LOC's https://github.com/recurly/callbacks/blob/c6edd0ce8e5d16bb759989794c60f60592a94214/api/callbacks/mime_test.go#L11

File: params_test.go [hunk 6]
```diff
@@ -20,3 +22,66 @@ func TestFormatTime(test *testing.T) {
...
+	t.Assert(ctx, c.Params.Context, "CommonParamsWithOptions()")
+}
+
+func TestWithIdempotencyKey(test *testing.T) {
+	t := &T{test}
+
+	key := "asdf"
+	p := &Params{}
+	opt := WithIdempotencyKey(key)
+	opt(p)
```
Comment (bmacnee) [inline params_test.go:61]: on these tests, would be worth checking the values that are set in p to make sure they are as-expected.

File: params_test.go [hunk 7]
```diff
@@ -20,3 +22,66 @@ func TestFormatTime(test *testing.T) {
...
+	t.Assert(key, p.IdempotencyKey, "WithIdempotencyKey()")
+}
+
+func TestWithContext(test *testing.T) {
+	t := &T{test}
+
+	ctx := context.Background()
+	p := &Params{}
+	opt := WithContext(ctx)
+	opt(p)
```
Comment (bmacnee) [inline params_test.go:72]: same here, lets confirm the ctx value of p was set correctly

File: params_test.go [hunk 8]
```diff
@@ -20,3 +22,66 @@ func TestFormatTime(test *testing.T) {
...
+}
+
+func TestWithHeader(test *testing.T) {
+	t := &T{test}
+
+	hKey := "Doug"
+	hVal := "Miller"
+	p := &Params{}
+	opt := WithHeader(http.Header{hKey: []string{hVal}})
+	opt(p)
```
Comment (bmacnee) [inline params_test.go:84]: same story

**Review by bmacnee — CHANGES_REQUESTED** [review]
A few initial comments, let's set up some time later today to discuss

File: client.go [hunk 9]
```diff
@@ -165,29 +165,30 @@ func BuildUrl(requestURL string, genericParams GenericParams) string {
 }
 
 // NewRequest generates an http.Request for the API client to submit to Recurly.
-func (c *Client) NewRequest(method string, requestURL string, params *Params) (*http.Request, error) {
+func (c *Client) NewRequest(method string, requestURL string, params *Params, requestOptions *RequestOptions) (*http.Request, error) {
```
Comment (bmacnee) [inline client.go:168]: I think it would be beneficial to write some comments on this function that address the oddities around params and requestOptions, and lay out the plan in the future to replace params

File: client_operations.go [hunk 10]
```diff
@@ -341,27 +341,29 @@ func (list *ListSitesParams) URLParams() []KeyValue {
 // API Documentation: https://developers.recurly.com/api/v2019-10-10#operation/list_sites
 //
 // Returns: A list of sites.
-func (c *Client) ListSites(params *ListSitesParams) *SiteList {
+func (c *Client) ListSites(params *ListSitesParams, opts ...Option) *SiteList {
 	path, err := c.InterpolatePath("/sites")
 	if err != nil {
 		// NOOP in 3.x client
 	}
+	requestOptions := NewRequestOptions(opts...)
```
Comment (bmacnee) [inline client_operations.go:349]: as discussed, we want to first apply the params in *listSitesParams, then apply the opts on top

File: params_test.go [hunk 11]
```diff
@@ -20,3 +20,67 @@ func TestFormatTime(test *testing.T) {
 	formattedTime := formatTime(time)
 	t.Assert(formattedTime, formatTime(time), "formatTime()")
 }
+
```
Comment (bmacnee) [inline params_test.go:23]: should we uncomment these so they're being tested?

**Review by bmacnee — COMMENTED** [review]
looking like a great foundation for 4.x, just a few comments

File: params_test.go (hunk 11 above)
Comment (douglasmiller) [reply params_test.go:23]: The code moved to `request_options.go`, but I haven't ported all of the tests over. I just commented these so that I can reference them later. They will be removed when ported over

File: request_options_test.go [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+			if tCase.idempotencyKey != "" {
+				opts = append(opts, WithIdempotencyKey(tCase.idempotencyKey))
+			}
+			if tCase.context != nil {
+				opts = append(opts, WithContext(tCase.context))
+			}
+			if tCase.header != nil {
+				opts = append(opts, WithHeader(tCase.header))
+			}
+			o = o.ApplyOptions(opts...)
```
Comment (bmacnee) [inline request_options_test.go:72]: we could make this test case more user-like
```go
		test.Run(tCase.name, func(test *testing.T) {
			t := &T{test}
			o := NewRequestOptions(WithIdempotencyKey(tCase.idempotencyKey), WithContext(tCase.context), WithHeader(tCase.header))
```

**Review by bmacnee — COMMENTED** [review]
couple final thoughts on unit testing so we get more coverage and some security

File: request_options_test.go [hunk 13]
```diff
@@ -0,0 +1,82 @@
+package recurly
```
[the remark this answers is not in the document]
Comment (bmacnee) [reply request_options_test.go:1]: keep in mind that there is an exception to the rule here. normally you can't have different packages in the same directory, but you can if the test package is specifically named the package name + "_test"

File: client_operations.go [hunk 14]
```diff
@@ -17,355 +17,15 @@ const (
...
-
-	ReactivateUniqueCouponCode(uniqueCouponCodeId string) (*UniqueCouponCode, error)
-
-	CreatePurchase(body *PurchaseCreate) (*InvoiceCollection, error)
-
-	PreviewPurchase(body *PurchaseCreate) (*InvoiceCollection, error)
-
-	GetExportDates() (*ExportDates, error)
-
-	GetExportFiles(exportDate string) (*ExportFiles, error)
```
Comment (bmacnee) [inline client_operations.go:272]: did you mean to delete these?

File: params.go [hunk 15]
```diff
@@ -30,6 +30,11 @@ type Params struct {
 	Data interface{} `json:"-"`
 }
 
+func (p *Params) clone() *Params {
+	copy := *p
```
Comment (bmacnee) [inline params.go:34]: you'll need to do a deeper copy here if you want p to be completely separate from copy. Params.Header is a map type, Params.Context, requestParams, and data are interfaces, so they will all be copied by their addresses, not by values.

Copying a map:
```go
for k,v := range originalMap {
  newMap[k] = v
}
```
interfaces are tougher. because you can't copy an interface, since its just a pointer to a struct. 
Let's schedule some time to talk about this

**Review by bmacnee — CHANGES_REQUESTED** [review]

File: request_options.go [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+		req.Header.Add("Idempotency-Key", o.IdempotencyKey)
+	}
+	for key, v := range o.Header {
+		for _, value := range v {
+			req.Header.Set(key, value)
+		}
+	}
+
+	if o.Context != nil {
+		req = req.WithContext(o.Context)
```
Comment (douglasmiller) [inline request_options.go:92]: Caught an issue with this creating a new request and we weren't returning it.

File: request_options.go [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+	Header http.Header `json:"-"`
+	// Context passed to the HTTP request for cancelling requests
+	Context context.Context `json:"-"`
+}
+
+func (o *RequestOptions) clone() *RequestOptions {
+	clone := &RequestOptions{}
+	clone.Header = o.Header.Clone()
+	clone.IdempotencyKey = o.IdempotencyKey
+	clone.Context = o.Context
```
Comment (bmacnee) [inline request_options.go:22]: since context is an interface, copying it here will copy a pointer to the same struct. that means if the requestOptions that is cloned cancels the context, the context will also be cancelled on the parent.

File: request_options.go [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+	o := &RequestOptions{}
+	return o.ApplyOptions(opts...)
+}
+
+// RequestOptionsFromParams will create a RequestOptions from a Params
+func RequestOptionsFromParams(p Params, opts ...Option) *RequestOptions {
+	o := &RequestOptions{}
+	o.Header = p.Header.Clone()
+	o.IdempotencyKey = p.IdempotencyKey
+	o.Context = p.Context
```
Comment (bmacnee) [inline request_options.go:41]: same concern here with copying

**Review by bmacnee — APPROVED** [review]
LGTM, let's get a review from someone else on the DX team in case im missing anything app-specific.

File: client.go [hunk 19]
```diff
@@ -137,7 +144,7 @@ func (c *Client) Call(method string, path string, genericParams GenericParams, v
 }
 
 // Append URL parameters
-func BuildUrl(requestURL string, genericParams GenericParams) string {
+func BuildURL(requestURL string, genericParams GenericParams) string {
```
Comment (chrissrogers) [inline client.go:147]: +1 for following the Go naming convention for acronyms

File: client.go [hunk 20]
```diff
@@ -165,29 +172,25 @@ func BuildUrl(requestURL string, genericParams GenericParams) string {
...
-	req.Header.Add("Accept", acceptVersion)
+	req = requestOptions.applyOptions(req)
+
+	// Set required headers after custom headers so that they aren't accidentally overwritten.
+	req.Header.Set("Accept", acceptVersion)
 	req.Header.Set("Accept-Encoding", "gzip")
-	req.Header.Add("Content-Type", "application/json; charset=utf-8")
-	req.Header.Add("User-Agent", userAgent)
+	req.Header.Set("Content-Type", "application/json; charset=utf-8")
+	req.Header.Set("User-Agent", userAgent)
```
Comment (chrissrogers) [inline client.go:190]: 👍

File: client_test.go [hunk 21]
```diff
@@ -171,3 +171,64 @@ func TestHeadResource200(test *testing.T) {
 		t.Errorf("resp.TotalRecords is incorrect. Expected: %v Got: %v", 100, intVal)
 	}
 }
+
+func TestSetCustomHeader(test *testing.T) {
+	t := &T{test}
+
+	hKey := "Doug"
+	hVal := "Miller"
```
Comment (chrissrogers) [inline client_test.go:179]: Nice guy

File: request_options.go [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+	applyOptions(req *http.Request) *http.Request
+}
+
+func (o *RequestOptions) applyOptions(req *http.Request) *http.Request {
+	for key, v := range o.Header {
+		for _, value := range v {
+			req.Header.Set(key, value)
+		}
+	}
+	// TODO: generate an idempotency key if missing?
```
Comment (chrissrogers) [inline request_options.go:74]: nit: we can probably remove this TODO since I don't think we'd want to generate idempotency keys in the client unless explicitly requested to do so.

**Review by chrissrogers — APPROVED** [review]
